### PR TITLE
Rewrite re2c tests runner in Python

### PR DESCRIPTION
**Pull request by sergeyklay**
This pull request introduces a completely rewritten test runner in Python.

The main goals of this change:
- Provide a OS-independent way to run tests (at least using CMake)
- Getting rid of dependence on Unix utilities in tests (grep, cat, tr, etc.)
- Flexibility
- Ease of maintenance

### CODE REVIEW

**Comment by sergeyklay** [comment]
> There is no `/home/skvadrik/devel/50/re2c/include` among them, and no suitable relative path (considering that the working directory is `/home/skvadrik/devel/50/re2c/test_210616222213/`, `-I include` means a copy of `/home/skvadrik/devel/50/re2c/test/include`, not `/home/skvadrik/devel/50/re2c/include`.
> 
> If I patch the command and add `-I ../include` then
> `{c,go}/unicode_identifier.re` do not fail anymore and I'm down to two
> failures:

Yes you are right. This is exactly what I decided at the very beginning. However, this doesn't work for some reason:

Right now I tried:
```diff
diff --git a/run_tests.py b/run_tests.py
index 9c93e4a9..2551e9ee 100644
--- a/run_tests.py
+++ b/run_tests.py
@@ -539,7 +539,8 @@ def init_process(base_path, skeleton=False, keep_temp_files=False,
     # abshere('@top_srcdir@', 'include')
     #   vs
     # abshere('..', '@top_srcdir@', 'include')
-    _ctx.incpaths.append(f"-I {abshere(base_path, 'include')}")
+    _ctx.incpaths.append("-I /Users/egrep/work/re2c/include")
+    _ctx.incpaths.append("-I ../include")
 
     # Set tools paths
     _ctx.diff = which('diff')[0]
```

And output is still the same:
```sh
$ python run_tests.py --keep-temp-files

-----------------
All:         1615
Ran:         1615
Passed:      1611
Soft errors: 0
Hard errors: 4
-----------------
FAILED
```

**Comment by sergeyklay** [comment]
@skvadrik Can you please commit your patch with the fix of the two tests?  I may be missing something

**Comment by skvadrik** [comment]
Thanks a lot for working on this!

One thing off the top of my head is that skeleton test failures can be deceptive: if re2c fails with an error, it is counted as a "soft error" because we have some tests that deliberately fail. Currently I see 222 soft failures in run_tests.py and 221 for run_tests.sh. The one extra error will likely go away after fixing non-skeleton tests, but the whole "soft error" concept is very weak and I added it as a temporary hack to monitor how many tests actually run and how many fail before that.

**Comment by skvadrik** [comment]
> @skvadrik Can you please commit your patch with the fix of the two tests? I may be missing something

I think I know what it is, it should be two separate arguments `-I` and `./include`, not `-I ../include`. At least I tried that change and the errors dropped from 4 to 2.

**Comment by sergeyklay** [comment]
> Currently I see 222 soft failures in run_tests.py and 221 for run_tests.sh.

Yeah, I saw this yesterday, but for now I decided to focus on non-skeleton tests.

**Comment by sergeyklay** [comment]
> I think I know what it is, it should be two separate arguments `-I` and `./include`, not `-I ../include`. At least I tried that change and the errors dropped from 4 to 2.

Yay! I spent about 8 hours on this! You're Chuck Norris :)

**Comment by skvadrik** [comment]
> Yay! I spent about 8 hours on this! You're Chuck Norris :)

I'm used to spending 8h a day on things like this (most of my working day really), so no wonder I could spot it. :D

**Comment by sergeyklay** [comment]
Yay!


> I think I know what it is, it should be two separate arguments `-I` and `./include`, not `-I ../include`. At least I tried that change and the errors dropped from 4 to 2.

Does this mean I should change this as well?
```diff
diff --git a/run_tests.py b/run_tests.py
index 8b84fa3d..ecd8ae57 100644
--- a/run_tests.py
+++ b/run_tests.py
@@ -532,7 +532,8 @@ def init_process(base_path, skeleton=False, keep_temp_files=False,
     # Set include paths, relative to build directory
     cwd = os.getcwd()
     os.chdir(_ctx.base_path)
-    _ctx.incpaths = [f'-I {d}' for d in Path('.').rglob('*') if os.path.isdir(d)]
+    paths = Path('.').rglob('*')
+    _ctx.incpaths = sum([['-I', str(d)] for d in paths if os.path.isdir(d)], [])
     os.chdir(cwd)
 
     # TODO: Fix this to sort out with absolute and relative paths + @top_srcdir@
```


```sh
-----------------
All:         1615
Ran:         1615
Passed:      1615
Soft errors: 0
Hard errors: 0
-----------------
```


Explanation

```python
list1 = ['111', '222', '333']
sum([['-I', item] for item in list1], [])

# Out: ['-I', '111', '-I', '222', '-I', '333']
```

**Comment by sergeyklay** [comment]
@skvadrik I updated the description of the PR. All tests are passed now.

**Comment by skvadrik** [comment]
> @skvadrik I updated the description of the PR. All tests are passed now.

\o/

**Comment by sergeyklay** [comment]
It is ready to review

**Comment by skvadrik** [comment]
> It is ready to review

Awesome, thanks! I may be unable to review it today, but I'll do it at the first opportunity.

**Comment by sergeyklay** [comment]
There are some issues with external tools on Windows. I'll try to sort out in next PRs to not mix changes in this one

**Comment by sergeyklay** [comment]
Some good (or bad) news: I managed to run tests runner under Windows. And I see half of the tests are failed. On the one hand, this is bad news, but on the other hand, we did not have the opportunity to run native compilation (not Mingw) and tests under WIndows until this moment. So now we see the real picture

**Comment by skvadrik** [comment]
> Some good (or bad) news: I managed to run tests runner under Windows. And I see half of the tests are failed. 

Can you start a new bug and post detail how the tests failed? E.g. upload the temporary test directory as an archive.

**Comment by sergeyklay** [comment]
> Can you start a new bug and post detail how the tests failed? E.g. upload the temporary test directory as an archive.

Sure. Let me some time to sort out with unstable test runner under the Windows (some regex compilation issues under the loops)

**Comment by sergeyklay** [comment]
> Some good (or bad) news: I managed to run tests runner under Windows. And I see half of the tests are failed. On the one hand, this is bad news, but on the other hand, we did not have the opportunity to run native compilation (not Mingw) and tests under WIndows until this moment. So now we see the real picture

I fixed tests runner to run under the Windows. There were some funny bugs like Windows path separator interpreted as a regex escape command, some Cmake issues related to build under the Windows and so on. And at the moment we have the following at GitHub actions:


```
Running in 2 job(s) 
----------------- 
All: 1615 
Ran: 1615 
Passed: 264 
Soft errors: 0
Hard errors: 1351 
----------------- 
FAILED
```

 I'll open a new PR with all necessary changes and show you all the remaining issues. But lets do it after this one will be merged so that I can rebase onto the proper branch and open a PR.

**Review by skvadrik — CHANGES_REQUESTED** [review]
Thanks again for doing this work! I don't know python very well, and to me the code looks nice and clean (much better than bash!). There is a couple of things to fix, but largely it just works. \o/

File: run_tests.py.in
Comment (skvadrik) [reply run_tests.py.in:627]: In case you want to rewrite it, there is some caution needed to avoid deleting too many or too few things. I check it by making some changes to the source code (e.g. change a couple of symbols in src/codegen/render.cc) to make tests fail in a "natural" way, and then make sure the number of reported failures is the same as `find . -name '*.re' | wc -l` in the test directory.

**Comment by skvadrik** [comment]
By the way, the new script is already ~2x faster than the old one. :)

File: run_tests.py.in
Comment (skvadrik) [inline run_tests.py.in:311]: "use un sort"  did you mean "use in sort"?

File: run_tests.py.in
Comment (trofi) [inline run_tests.py.in:603]: Shell script did a simplistic static chunking just because it's not very convenient to implement more complex pool of workers. That usually introduces bigger imbalace between chunks than single input queue thrown at a worker pool.

In python variant I suggest just using one test per work item. The following seems to speed up tests from `1.5s` to `1.2s` on 16-CPU x86_64-linux machine:

```diff
--- a/run_tests.py.in
+++ b/run_tests.py.in
@@ -600,7 +600,7 @@ def main():
             initargs=(test_blddir, args.skeleton, args.keep_temp_files,
                       args.valgrind, args.wine,)
     ) as pool:
-        pool.map(run_pack, chunks(tests, int(len(tests) / jobs + 1)))
+        pool.map(run_pack, chunks(tests, 1))

     # collect results from all threads
     total = len(tests)
```

The caveat is that stats are somehow broken (though all tests we ran for real):

```
-----------------
All:         1605
Ran:         16
Passed:      16
Soft errors: 0
Hard errors: 0
-----------------
```

File: CMakeLists.txt
Comment (sergeyklay) [reply CMakeLists.txt:101]: I not familiar with this technique in autotools. Could you please send a commit to this branch?

File: run_tests.py.in
Comment (sergeyklay) [reply run_tests.py.in:603]: I did this initially :) But after I saw that the reports were broken, I adjusted the behavior to the bash implementation

File: CMakeLists.txt
Comment (skvadrik) [reply CMakeLists.txt:101]: In Autoconf that would be something like `AC_CHECK_PROG(python, ...)`. But now I thought better of it and decided not to do it, because I still want re2c to build even if there is no python (or python3). The user may not be able to run tests, but this may be ok (e.g. if re2c is a dependency that they have to build on some weird platform).

File: run_tests.py.in
Comment (skvadrik) [reply run_tests.py.in:603]: Let's merge the current version as is (with chunk size as in the old bash script), and then fix the stats and adjust chunk size in a follow-up change.

File: run_tests.py.in
Comment (skvadrik) [reply run_tests.py.in:627]: I sent you a PR https://github.com/sergeyklay/re2c/pull/2 with this commit --- let me know if there is some easier way to add commits on top of an existing PR.

Comment (sergeyklay) [reply run_tests.py.in]: @skvadrik I just changed this (a minute before your PR :) )

**Comment by sergeyklay** [comment]
@skvadrik It is ready to merge now.

**Review by skvadrik — APPROVED** [review]

**Comment by skvadrik** [comment]
Great, I have just merged it! \o/

**Comment by sergeyklay** [comment]
> Great, I have just merged it! \o/

Yay! I'm glad this finely happened :)

File: CMakeLists.txt
Comment (sergeyklay) [reply CMakeLists.txt:101]: This line still bothers me.  I think package maintainers who use CMake may have unexpected issues

Comment (skvadrik) [reply CMakeLists.txt:101]: I think it is fine --- a system that has CMake is likely to have python as well. The other option would be to drop REQUIRED and add a condition if (PYTHON3_FOUND) or some such when using python.

Autotools is a different story, because one can just unpack the release tarball (without the need to regenerate the configure script), and avoid any dependencies except for shell and C++ compiler. Not that I expect re2c to be used on systems without python, but I don't see any use in enforcing it either.

Comment (skvadrik) [reply CMakeLists.txt:101]: (sorry, I accidentally edited your comment and then restored it back)

**Pull request merged**
